### PR TITLE
11 implement user profile edit page

### LINKED ISSUE #11: Implement user profile edit page
User profile page that lets user edit:
○ Username
○ Name
○ Email
○ Profile photo
○ Bio
○ Password

### PR DESCRIPTION
**Pull request by dwilliamson5**
Questions:

- I couldn't remember if i made the views /users/edit and /users/profile. Do we want to use that instead of /profile/edit?
- I didn't include the password in the edit profile page, should it be included?
- Edit profile link ok on landing page?

### CODE REVIEW

**Comment by matttarantino** [comment]
1) I made them as stubs, but I think what you have makes sense so 👍 on that 
2) yeah, I think we need those fields (password + confirm password) (we will need validation to make sure they are the same value in both) (we shouldn't/can't populate the password field with the current value since its already encrypted) 
3) yep, thats perfect! 

One question I had:
- Should we let you change your university? Im thinking once you pick one, you are locked into that university and cannot edit it afterwards. Otherwise you would also need to edit your email address to match the new domain etc. Thoughts?

**Comment by dwilliamson5** [comment]
> 2. yeah, I think we need those fields (password + confirm password) (we will need validation to make sure they are the same value in both) (we shouldn't/can't populate the password field with the current value since its already encrypted)

I'll add this change in.

>     * Should we let you change your university? Im thinking once you pick one, you are locked into that university and cannot edit it afterwards. Otherwise you would also need to edit your email address to match the new domain etc. Thoughts?

I could go either way on this. If you attended a different university for undergrad vs graduate but you want to keep the same account? I do check if the new email matches the university. It's easy to delete so let me know if it should be removed.

**Comment by matttarantino** [comment]
I think there could be some cool JS stuff we would do (like throwing a warning if your selected university doesn't match your email domain letting the user know they need to match if they want to switch) however I worry it could cause a rabbit hole of other problems (like what if you have pending items for sale with bids and whatnot, and now we removed you from accessing those since you now attend a different university). Maybe we consider doing it if we have extra time at the end?

**Comment by dwilliamson5** [comment]
> I think there could be some cool JS stuff we would do (like throwing a warning if your selected university doesn't match your email domain letting the user know they need to match if they want to switch) however I worry it could cause a rabbit hole of other problems (like what if you have pending items for sale with bids and whatnot, and now we removed you from accessing those since you now attend a different university). Maybe we consider doing it if we have extra time at the end?

Ok that's fine i'll remove it

**Comment by dwilliamson5** [comment]
- Removed universities from edit form
- Added password to edit form
- Check that new username isn't already in use

File: views/profile/edit.handlebars [hunk 1]
```diff
@@ -0,0 +1,46 @@
+<p class="error-status">{{error_status_code}}</p>
+<p class="error-message">{{error_messages}}</p>
+
+<form action="/profile/edit" method="POST" id="signup-form">
```
Comment (matttarantino) [inline views/profile/edit.handlebars:4]: This is should a PUT right? We are updating an existing object

File: routes/profile.js [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                    error_messages: 'Internal Server Error',
+                    user: user
+                });
+                return;
+            }
+        }
+
+        res.redirect('/');
+    } catch (e) {
+        res.status(500).render('errors/500', {
```
Comment (matttarantino) [inline routes/profile.js:129]: We could probably render the 500 error on the profile/edit page just like we do on auth pages, right?

Comment (matttarantino) [reply routes/profile.js:129]: Just like on line 117

File: routes/profile.js [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        });
+        return;
+    }
+});
+
+router.post('/edit/', async (req, res) => {
+    let body = req.body;
+
+    let shouldUpdatePassword = false;
+    const user = await users.getUser(req.session.user.username);
```
Comment (matttarantino) [inline routes/profile.js:41]: this method can `throw` and we dont have a try/catch here

**Review by matttarantino — COMMENTED** [review]
Still working through reviewing this, but wanted to leave comments so far

**Comment by matttarantino** [comment]
@dwilliamson5 for the merge conflict, can we move the profile link to the header here? https://github.com/dwilliamson5/CS546_WS_final_project/blob/main/views/layouts/main.handlebars#L24

**Comment by matttarantino** [comment]
@dwilliamson5  going to close this since its already been merged in another PR

**Pull request closed without merging**